feat: pipeline-level quantization config

sayakpaul opened the pull request:
# What does this PR do?

See: https://github.com/huggingface/diffusers/issues/10327

TL;DR: This PR adds support to apply a quantization config when doing `DiffusionPipeline.from_pretrained(...)`, thereby making it easier for the users to benefit from quantization. 

## Why

To apply quantization to a `DiffusionPipeline`, a user has to first initialize the models they want to quantize with desired `quantization_configs`:

```py
quant_config = TransformersBitsAndBytesConfig(load_in_8bit=True,)

text_encoder_2_8bit = T5EncoderModel.from_pretrained(
    "black-forest-labs/FLUX.1-dev",
    subfolder="text_encoder_2",
    quantization_config=quant_config,
    torch_dtype=torch.float16,
)

quant_config = DiffusersBitsAndBytesConfig(load_in_8bit=True,)
transformer_8bit = FluxTransformer2DModel.from_pretrained(
    "black-forest-labs/FLUX.1-dev",
    subfolder="transformer",
    quantization_config=quant_config,
    torch_dtype=torch.float16,
)

pipe = DiffusionPipeline.from_pretrained(
    ..., transformer=transformer_8bit, text_encoder_2=text_encoder_2_8bit
)
```

This is cumbersome. 

## What

@SunMarc and I worked on this PR to show the kind of simple changes we need to enable a user to pass a quantization config directly whilst doing `DiffusionPipeline.from_pretrained(..., quantization_config=...)`. The user experience now becomes:

```py

pipeline_quant_config = PipelineQuantizationConfig(
    quant_backend="bitsandbytes_8bit",
    quant_kwargs={"load_in_8bit": True},
    components_to_quantize=["text_encoder_2", "transformer"]
)
pipe = DiffusionPipeline.from_pretrained(
    ...,
    quantization_config=pipeline_quant_config
)
```

Users can specify granular level quantization mapping too:

```py
quant_mapping = {
    "transformer": DiffBitsAndBytesConfig(
        load_in_4bit=True, bnb_4bit_quant_type="nf4", bnb_4bit_compute_dtype=torch.bfloat16
     ),
    "text_encoder_2": TranBitsAndBytesConfig(
        load_in_4bit=True, bnb_4bit_quant_type="nf4", bnb_4bit_compute_dtype=torch.bfloat16
     ),
}
pipeline_quant_config = PipelineQuantizationConfig(quant_mapping=quant_mapping)
```

This is particularly helpful when using different quantization backends for different modules (below we show a combination of Quanto and BitsAndBytes):

```py
quant_mapping = {
    "transformer": QuantoConfig(weights_dtype="float8"),
    "text_encoder_2": TranBitsAndBytesConfig(
        load_in_4bit=True, bnb_4bit_quant_type="nf4", bnb_4bit_compute_dtype=torch.bfloat16
     ),
}
pipeline_quant_config = PipelineQuantizationConfig(quant_mapping=quant_mapping)
```

Here's a script that might be helpful for others to test:

<details>
<summary>code</summary>

```py
from diffusers.quantizers import PipelineQuantizationConfig
from diffusers import DiffusionPipeline
import argparse
import torch


def get_global_config():
    quant_config = PipelineQuantizationConfig(
        quant_backend="bitsandbytes_4bit",
        quant_kwargs={"load_in_4bit": True, "bnb_4bit_quant_type": "nf4", "bnb_4bit_compute_dtype": torch.bfloat16},
        components_to_quantize=["transformer", "text_encoder_2"],
    )
    return quant_config


def get_granular_config(use_quanto=False):
    from diffusers import BitsAndBytesConfig as DiffBitsAndBytesConfig, QuantoConfig
    from transformers import BitsAndBytesConfig as TranBitsAndBytesConfig

    transformer_config = (
        QuantoConfig(weights_dtype="float8")
        if use_quanto
        else DiffBitsAndBytesConfig(
            load_in_4bit=True, bnb_4bit_quant_type="nf4", bnb_4bit_compute_dtype=torch.bfloat16
        )
    )

    quant_mapping = {
        "transformer": transformer_config,
        "text_encoder_2": TranBitsAndBytesConfig(
            load_in_4bit=True, bnb_4bit_quant_type="nf4", bnb_4bit_compute_dtype=torch.bfloat16
        ),
    }
    quant_config = PipelineQuantizationConfig(quant_mapping=quant_mapping)
    return quant_config


def load_pipeline(quant_config):
    pipe = DiffusionPipeline.from_pretrained(
        "black-forest-labs/FLUX.1-dev", quantization_config=quant_config, torch_dtype=torch.bfloat16
    ).to("cuda")
    return pipe


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--use_global_config", action="store_true")
    parser.add_argument("--use_quanto", action="store_true")
    args = parser.parse_args()

    quant_config = get_global_config() if args.use_global_config else get_granular_config(args.use_quanto)
    pipe = load_pipeline(quant_config)

    pipe_kwargs = {
        "prompt": "A cat holding a sign that says hello world",
        "height": 1024,
        "width": 1024,
        "guidance_scale": 3.5,
        "num_inference_steps": 50,
        "max_sequence_length": 512,
    }

    image = pipe(**pipe_kwargs, generator=torch.manual_seed(0)).images[0]
    image.save(f"quant_global@{args.use_global_config}_quanto@{args.use_quanto}.png")
```

</details>

Cc: @asomoza if you want to test this out :) 

## TODOs

- [x] Docs
- [x] Tests
- [x] Enable testing in CI

sayakpaul commented on src/diffusers/quantizers/__init__.py:
```diff
@@ -12,5 +12,165 @@
+        self,
+        quant_backend: str = None,
+        quant_kwargs: Dict[str, Union[str, float, int, dict]] = None,
+        modules_to_quantize: Optional[List[str]] = None,
```
Should there be a reasonable default for this? @SunMarc had some ideas around this.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_11130). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

SunMarc reviewed (commented):
I think that we can go ahead and start adding simple test and a bit of documentation !

sayakpaul commented:
@DN6 it would be great if you could do a first review of this

SunMarc commented:
cc @DerekLiu35 might be interesting to you !

yiyixuxu reviewed (commented):
the change to pipelines code looks good to me! 

the provided code example isn't working though (does not throw an error but no effect) - do we intend to support passing as a dict?

DN6 commented on src/diffusers/quantizers/__init__.py:
```diff
@@ -12,5 +12,165 @@
+        self,
+        quant_backend: str = None,
+        quant_kwargs: Dict[str, Union[str, float, int, dict]] = None,
+        modules_to_quantize: Optional[List[str]] = None,
```
Nit. Would call this `components_to_quantize`

sayakpaul replied:
Done.

DN6 commented on src/diffusers/quantizers/__init__.py:
```diff
@@ -12,5 +12,165 @@
+
+        if init_kwargs_transformers != init_kwargs_diffusers:
+            raise ValueError(
+                "The signatures of the __init__ methods of the quantization config classes in `diffusers` and `transformers` don't match. "
```
Would just provide an example snippet for what the mapping looks like.

sayakpaul replied:
Done.

DN6 reviewed (commented):
Looking good. Could we add a test as well please.

DN6 commented on src/diffusers/quantizers/__init__.py:
```diff
@@ -12,5 +12,165 @@
+            msg = ""
+            if not (any(isinstance(config, available) for available in available_configs_diffusers)):
+                msg = f"Provided config for {module_name=} could not be found. Available ones for `diffusers` are: {available_configs_diffusers}.)"
+            elif available_configs_transformers is not None and not (
```
This line throws an error when running this snippet

```python
import torch
from diffusers import DiffusionPipeline
from diffusers.quantizers.quantization_config import QuantoConfig
from diffusers.quantizers import PipelineQuantizationConfig
from transformers import BitsAndBytesConfig

pipeline_quant_config = PipelineQuantizationConfig(
    quant_mapping={
        "transformer": QuantoConfig(weights_dtype="int8"),
        "text_encoder_2": BitsAndBytesConfig(
            load_in_4bit=True, compute_dtype=torch.bfloat16
        ),
    }
)
pipe = DiffusionPipeline.from_pretrained(
    "black-forest-labs/FLUX.1-dev",
    quantization_config=pipeline_quant_config,
    torch_dtype=torch.bfloat16,
)
```

```
Traceback (most recent call last):
  File "/home/dhruv/diffusers/../scripts/test_pipeline_quant.py", line 13, in <module>
    pipeline_quant_config = PipelineQuantizationConfig(
  File "/home/dhruv/diffusers/src/diffusers/quantizers/__init__.py", line 51, in __init__
    self.post_init()
  File "/home/dhruv/diffusers/src/diffusers/quantizers/__init__.py", line 57, in post_init
    self._validate_init_args()
  File "/home/dhruv/diffusers/src/diffusers/quantizers/__init__.py", line 73, in _validate_init_args
    self._validate_quant_mapping_args()
  File "/home/dhruv/diffusers/src/diffusers/quantizers/__init__.py", line 116, in _validate_quant_mapping_args
    raise ValueError(msg)
ValueError: Provided config for module_name='transformer' could not be found. Available ones for `transformers` are: [<class 'transformers.utils.quantization_config.AwqConfig'>, <class 'transformers.utils.quantization_config.BitsAndBytesConfig'>, <class 'transformers.utils.quantization_config.BitsAndBytesConfig'>, <class 'transformers.utils.quantization_config.EetqConfig'>, <class 'transformers.utils.quantization_config.GPTQConfig'>, <class 'transformers.utils.quantization_config.AqlmConfig'>, <class 'transformers.utils.quantization_config.QuantoConfig'>, <class 'transformers.utils.quantization_config.QuarkConfig'>, <class 'transformers.utils.quantization_config.HqqConfig'>, <class 'transformers.utils.quantization_config.CompressedTensorsConfig'>, <class 'transformers.utils.quantization_config.FbgemmFp8Config'>, <class 'transformers.utils.quantization_config.HiggsConfig'>, <class 'transformers.utils.quantization_config.TorchAoConfig'>, <class 'transformers.utils.quantization_config.BitNetConfig'>, <class 'transformers.utils.quantization_config.VptqConfig'>, <class 'transformers.utils.quantization_config.SpQRConfig'>, <class 'transformers.utils.quantization_config.FineGrainedFP8Config'>].)
```

sayakpaul replied:
Done fixing this. Thanks for catching @DN6

sayakpaul commented:
@DN6 @yiyixuxu @SunMarc thanks for the feedback. 

I have added a test suite to check for the most crucial stuff. 

@yiyixuxu,

> the provided code example isn't working though (does not throw an error but no effect) - do we intend to support passing as a dict?

Maybe there were bugs that were untested but they should be fixed now. For example, the following works for me fine:

<details>
<summary>Expand</summary>

```py
import torch
from diffusers import DiffusionPipeline
from diffusers.quantizers.quantization_config import QuantoConfig
from diffusers.quantizers import PipelineQuantizationConfig
from transformers import BitsAndBytesConfig

torch.cuda.empty_cache()
torch.cuda.reset_max_memory_allocated()

print(f"Memory: {torch.cuda.max_memory_allocated() / (1024 ** 3)} GB.")

pipeline_quant_config = PipelineQuantizationConfig(
    quant_backend="bitsandbytes_4bit",
    quant_kwargs={"load_in_4bit": True, "bnb_4bit_quant_type": "nf4", "bnb_4bit_compute_dtype": torch.bfloat16},
    components_to_quantize=["transformer", "text_encoder_2"],
)
pipe = DiffusionPipeline.from_pretrained(
    "black-forest-labs/FLUX.1-dev",
    quantization_config=pipeline_quant_config,
    torch_dtype=torch.bfloat16,
).to("cuda")

for name, component in pipe.components.items():
    if isinstance(component, torch.nn.Module):
        if getattr(component.config, "quantization_config", None) is not None:
            print(name, component.config.quantization_config)

pipe("dog", num_inference_steps=5)

print(f"Memory: {torch.cuda.max_memory_allocated() / (1024 ** 3)} GB.")
```

</details>

Additionally, the code uses about 15 GB of memory, whereas without quantization, we get about 34 GB. So, I think it's working as expected? 

Regarding passing as dicts, I like the idea however my stance is that passing inputs through a Config class just allows us to perform quick validations and keep that code out of the score `pipeline_utils.py` stuff. LMK your thoughts.

sayakpaul commented:
@stevhliu where do you think we should document it?

stevhliu commented:
We can document the two ways to pass quantization configs directly on the [overview](https://huggingface.co/docs/diffusers/quantization/overview) or we can create a new doc and add it after the overview.

I think adding directly to the overview would probably be the easiest.

It would also be nice to recommend when to use each method. With this new `PipelineQuantizationConfig` (which seems much simpler), are there still scenarios where a user would want to initialize the model first and then pass a quantization config?

sayakpaul commented:
> We can document the two ways to pass quantization configs directly on the [overview](https://huggingface.co/docs/diffusers/quantization/overview) or we can create a new doc and add it after the overview.

> I think adding directly to the overview would probably be the easiest.

I feel the same. Done in 872c91ef6.

> With this new PipelineQuantizationConfig (which seems much simpler), are there still scenarios where a user would want to initialize the model first and then pass a quantization config?

Good question. I don't think there would be any need. We will see in due time.

yiyixuxu commented:
@sayakpaul 
I'm not suggesting passing as a dict, I made that comment because in the code you provided for test in the PR description, you passed it as a dict and the pipeline does not throw an error 


```python
def get_granular_config(use_quanto=False):
    from diffusers import BitsAndBytesConfig as DiffBitsAndBytesConfig, QuantoConfig
    from transformers import BitsAndBytesConfig as TranBitsAndBytesConfig

    transformer_config = (
        QuantoConfig(weights_dtype="float8")
        if use_quanto
        else DiffBitsAndBytesConfig(
            load_in_4bit=True, bnb_4bit_quant_type="nf4", bnb_4bit_compute_dtype=torch.bfloat16
        )
    )

    quant_config = {
        "transformer": transformer_config,
        "text_encoder_2": TranBitsAndBytesConfig(
            load_in_4bit=True, bnb_4bit_quant_type="nf4", bnb_4bit_compute_dtype=torch.bfloat16
        ),
    }
    return quant_config
```

sayakpaul commented:
@yiyixuxu my bad. I have updated the code snippet. I have also added a check for validating the `quantization_config` input to `DiffusionPipeline` (and a test for it).

SunMarc commented on docs/source/en/quantization/overview.md:
```diff
@@ -39,3 +39,79 @@ Diffusers currently supports the following quantization methods.
+pipeline_quant_config = PipelineQuantizationConfig(
+    quant_mapping={
+        "transformer": QuantoConfig(weights_dtype="int8"),
+        "text_encoder_2": BitsAndBytesConfig(
```
One caveat of this method is that they need to know which component comes from transformers because they need to import the right configuration. Maybe we can mention this somewhere ?

SunMarc commented on docs/source/en/quantization/overview.md:
````diff
@@ -39,3 +39,79 @@ Diffusers currently supports the following quantization methods.
+
+```py
+pipeline_quant_config = PipelineQuantizationConfig(
+    quant_backend="bitsandbytes_4bit",
````
Can you share a list of the quant_backend values ?

SunMarc reviewed (approved):
Thanks a lot ! LGTM !

DN6 commented on src/diffusers/quantizers/__init__.py:
```diff
@@ -12,5 +12,183 @@
+            raise ValueError(
+                "The signatures of the __init__ methods of the quantization config classes in `diffusers` and `transformers` don't match. "
+                f"Please provide a `quant_mapping` instead, in the {self.__class__.__name__} class. Refer to the docs to learn more about how "
+                "this mapping would look like: TODO."
```
Snippet needs to be included here no or the link to the docs? Is the intent to add this in after merge?

DN6 commented on .github/workflows/nightly_tests.yml:
```diff
@@ -525,6 +525,60 @@ jobs:
+        run: |
+          python -m venv /opt/venv && export PATH="/opt/venv/bin:$PATH"
+          python -m uv pip install -e [quality,test]
+          python -m uv pip install -U bitsandbytes optimum_quanto
```
TorchAO as well no?

sayakpaul replied:
For the CI, we're testing for bnb and quanto.

DN6 reviewed (approved):
Minor comments that are not merge blocks. Can be addressed here or in a follow up.

sayakpaul commented:
Thanks all for your comments! Cc: @asomoza @apolinario @linoytsaban for awareness of this feature as it makes the barrier to entry for quantization of a `DiffusionPipeline` lots easier.

nitinmukesh commented:
I just came across this implementation while refering to Chroma example. Very good feature, thank you @sayakpaul
